7275: manage translation schemas proposal (#10078)
CDCgov/prime-reportstream: prime-router/docs/features/0005-managing-translation-schemas.md

# Managing Translation Schemas

## Context

Translation schemas are used at multiple points in the universal pipeline

> - transforming items sent by senders

Reviewer: ```suggestion
- transforming items sent by senders from their chosen format to ReportStream's internal FHIR format
```
Consider this change, if you want

> - transforming items right before they are delivered to a receiver

Reviewer: ```suggestion
- transforming items from ReportStream's internal FHIR format to the chosen format of the receiver right before sending
```
Consider this change, if you want

and cover different kinds of transforms:

- HL7 -> FHIR
    - **This is handled by another [library](https://github.com/LinuxForHealth/hl7v2-fhir-converter) with limited configuration option of where to read the mapping files from**
- FHIR -> FHIR
- FHIR -> HL7

These translation schemas are currently stored as files that are included in the deployed application and read from the file sytem
when they are applied to an item.

## Problem

The overarching problem is that since the translation schemas are part of the deployed application they can only be updated at the frequency at which the application itself is deployed.  This is quite limiting and has downstream impacts such as:

- onboarding new senders/receivers can take a while as iterative changes to the translation schema have to be tested out over several days or weeks
- bugs discovered in the translation schema cannot be immediately addressed without a hotfix

> - the senders/receivers do not have any capability to self-serve changes to their own translation schemas

Reviewer: I think they still won't with this design, so we should make that clear somewhere below or remove this? I would vote removing it because we have determined this isn't something we want to support anyways...

Author: Yep I'll drop

## Goal

> Design a new implementation that enables iterating on the translation schemas outside of the deployments and unlocks the ability to implement a feature that enables sender/receivers to self-serve on changes.

Reviewer: Again, this goal should be reworded a bit since users self serving isn't a goal of ours

- A custom solution would need to be implemented for the HL7->FHIR since the library only reads the mapping from disk
- Testing this change would be extremely complicated especially considering the current infrastructure around testing the DB
- Would make it much more difficult to extract libraries for other users to consume as the libraries would be more closely tied to the db

### Azure blob storage (Recommended)
This solution would maintain the majority of the existing functionality and just provide updates to use the azure blob storage as the underlying "file" storage rather than disk.  This solution would get versioning automatically by enabling [azure blob versioning](https://learn.microsoft.com/en-us/azure/storage/blobs/versioning-overview); this functionality enables automatically creating new versions when uploading a new file and then APIs for rolling back file updates in the event that there was a problem.

Additionally, the versioning has the added flexibility of referencing a specific version of a blob so translation schemas could reference version 3 while the next version is iterated on.

> `https://<storage-account>.blob.core.windows.net/<container>/<blob-name>?versionid=<version-id>` . If the decision is made to continue to store some of the common translation schemas in source code this is an approach that could be adopted here as well by adding a version to the directory paths `file:///{BASE_DIR}/metadata/hl7_mapping/common/patient/v1/patient.yml``

Reviewer: Regarding "If the decision is made..." I think we should make these decisions now and have the document reflect them, otherwise it just gets kind of burdensome.

Reviewer: My opinion is we should go with the idea that "core" mappings/translations will stay in the repo and only "end user" translations will be stored in the blob store as these are the ones that often see the most tweaks and changes. Once we create mappings for an ORU-R01 message, for example, those ideally should never change...

Author: My thought was to keep it open ended and aim for a solution that supports all above.  For example the ORU-R01 mapping might change if the RADxMARs spec change.

Reviewer 2: If the RADxMARs spec changes, I think that should be handled by a RADxMARs transform and keep ORU-R01 un-touched.

Author: Yep, my comment was more oriented around that the RADxMARS mappings for ORU_01 might change (since they live in the same directory)

#### FHIR -> FHIR, FHIR -> HL7

> Steps for migrating:

Reviewer: This seems to be written from the perspective of ALL the mappings moving to the blob store. If we decide to keep core mappings in the repo, then we should update this section.

Author: So all the schemas need to get updated to no longer have relative pathing, but I'll update step 3 to make it clear that only receiver/sender translation schemas will be added

#### Cache
**This could potentially be deferred until it's been identified that the network requests to azure blob storage are the bottleneck**

Since the pipeline will now require reading several files out of the blob store at multiple spots, it will likely make sense to cache the resolved translation schemas since they will change relatively infrequently.  This could initially be done with a simple in-memory cache using the translation schema URI as the cache key.

#### Pros
- Maintains the current file based model for the translation schemas
- Translation schemas can be easily edited by hand via the Azure UI
- Supports reading translation schemas from azure as well as from disk
    - Enables leaving the common translation schemas on disk if that's preferred
- Enables easy rollback
    - Post MVP, a UI could be enabled to allow selecting what version to go back to
- Supports reading files from azure or from disk

#### Cons

> - Need to support to different URI schemes

Reviewer: ```suggestion
- Need to support different URI schemes
```

- Relies on pulling files out of azure which is slower than disk or the database
- Requires more complicated error handling

## Managing translation schemas
The translation schemas break down into two categories:

- translation schemas dedicated to a specific sender or receiver
  - one of them will typically be set in the `setting` for either the sender or receiver
- translation schemas that are used across multiple senders or receivers across organizations

>     - for now, these will continue to be stored in the source code since changes here should be carefully rolled out

Reviewer: ```suggestion
    - for now, these will continue to be stored in the source code since changes here should be carefully rolled out and they are not subject to frequent changes
```

    - an example of this would be `ADT-01` directory

### Directory Structure

Below is the directory structure that we would implement inside the azure blob store.  It includes the location of

> where common translation schemas used across organizations would go, but the current thought is whether (and which ones) they should exist in the blob

Reviewer: ```suggestion
where common translation schemas used across organizations would go, but the current design states these should remain in the repository.
```

> store will be decided down the road since this design supports the file living in either location.  The largest change from the existing

Reviewer: ```suggestion
The largest change from the existing
```

directory structure is that a sample input and output file will be included in order to verify the translation schema works as expected.

**Note: the `fhir_mapping` directory will be lifted exactly as is to azure since the directory structure is decided by the underlying library**

```
/azure-blob-container
    /fhir_mapping

>         valid-2023-03-29T23:53:00.488Z.txt

Reviewer: This is confusing to me, each schema should have its own input/output test file, no? OR at least its own output test file?

Author: Each one does... but because of `extend` our validity exists for all the schemas collectively, not an individual basis.

also used in an `extend` property. See [default-sender-transform.yml](https://github.com/CDCgov/prime-reportstream/blob/3e69f874c78d49bd69b8303a38ba3e0a747ef3f9/prime-router/metadata/fhir_transforms/senders/default-sender-transform.yml)
and [simple-report-sender-transform.yml](https://github.com/CDCgov/prime-reportstream/blob/3e69f874c78d49bd69b8303a38ba3e0a747ef3f9/prime-router/metadata/fhir_transforms/senders/SimpleReport/simple-report-sender-transform.yml)
as an example.

A side effect of this is that translation schemas cannot be atomically edited and validated.  Using the above example, changing `default-sender-transform`
would necessitate at the minimum changing the sample out for `simple-report-sender-transform` and possibly require making a change
to the translation schema itself.  This issue can end up involving many translation schemas as there are no limits on the use of the `extend` property.

The workflow that will be employed to create or update translation schemas will be the following.

1. Developer syncs all the translation schemas in azure to their local environment
2. Developer makes changes to the translation schemas and their samples

> 3. Developer verifies these changes by running the validation that asserts that running the translation schema with its sample input matches the sample output

Author: This was meant to be a high level overview, are you clear after the other sections?

4. Developer syncs translation schemas to the staging azure container and a `validating.txt` file is created
5. An azure function with a BlobTrigger on the existence of `validating.txt` and runs the same validation in staging
    - If the validation fails a page goes off and all files are rolled back to a previous version
6. The `validating.txt` file is replaced with a `valid-{timestamp}.txt`

> 7. The developer runs the GitHub action that syncs the staging and production azure containers performing the same validation

Author: Addressed in a new section on handling schemas still in source code

      - the current `valid-{timestamp}.txt` is renamed `previous-valid-{timestamp}.txt` 
      - invoke the [`azcopy`](https://learn.microsoft.com/en-us/azure/storage/common/storage-ref-azcopy-copy)
    command
          - `azcopy cp '{from}' '{to}' --recursive`
      - create a `validating.txt` file
  - The special case for this command will be copying translation schemas from staging/production to the localhost which will not perform any validation
  - Arguments
      - `transformType`
          - Options: `hl7-fhir`, `fhir-fhir`, `fhir-hl7`
      - `from`
          - Options: `local`, `staging`, `production`
      - `to`
         - Options: `local`, `staging`, `production`
- `validateSchemas`
  - This command will validate that all the translation schemas work as expected by iterating over each of them and executing them against the sample input and verifying
  that it matches the sample output and will be used as part of the local development flow.
  - Arguments
     - `transformType`
        - Options: `hl7-fhir`, `fhir-fhir`, `fhir-hl7`

New Azure Functions:

> - `Validate{TransformType}Schemas`

Author: Yea under the hood the code will be the same just need separate functions/triggers

  - This will be triggered upon the creation of a `validating.txt`
  - The function will perform the following actions
    - Run all the translation schemas against their sample inputs and validate that it matches the sample output
      - If this validation fails, a page will be generated and all the files will be rolled back to their previous version
    - Replace the `validating.txt` with a `valid-{timestamp}.txt`
      - If the validation fails, the `previous-valid-{timestamp}.txt` will simply be renamed to `valid-{timestamp}.txt`
      - If the validation passes, a new `valid-{timestamp}.txt` is created

New GitHub action:

- Sync Staging Schemas to Production
  - In order to avoid having developers keep azure blob credentials locally as well as to have a better access log of when
  translation schema updates are applied to production, this new GitHub action will be used to sync translation schemas between staging and production
  - It will use the same workflow of invoking the `syncSchemas` command and will then poll for an updated `valid-{timestamp}.txt`

Example of the flow with a sample directories that shows adding a new translation schema:

> 1. Developer picks up ticket to create a new translation schema for Flexion

Reviewer: ```suggestion
1. Developer picks up ticket to create a new translation schema for Trusted Intermediary 
```

                    previous-valid-2023-03-29T23:53:00.488Z.txt
                    /default
                        input.fhir
                        output.fhir
                        default-sender-transform.yml
                    /Flexion
                        flexion-sender-transform.yml
                        input.fhir
                        output.fhir
        ```
    </details>
    <details>
      <summary>Production</summary>

        ```
            /local-azure-container
                /fhir_transforms
                    valid-2023-04-03T23:53:00.488Z.txt
                    previous-valid-2023-03-29T23:53:00.488Z.txt
                    /default
                        input.fhir
                        output.fhir
                        default-sender-transform.yml
                    /Flexion
                        flexion-sender-transform.yml
                        input.fhir
                        output.fhir
        ```
    </details>

### Validation

The initial validation will mimic how the current integration tests work.  Each translation schema will:

- be checked to make sure it can be successfully

> - ran with its sample input and checked against its sample output

Reviewer: combine this bullet with the above?

  - failing translation schemas will be indicated in the output

Some follow-up improvements could include indicating which lines did not match between the sample output and produced value
or what line was invalid in the schema.

## Operations

### Rollout

1. Implement the code update to support using the URI scheme to determine how to retrieve the file
2. Update all current translation schemas in source code to use the `file:///` scheme
3. Create the workflow for creating/updating a translation schema used directly in the convert or translate step
    - This would be a translation schema that is referenced in a `setting`
4. Migrate all translation schemas referenced in a `setting` to azure
5. Evaluate whether schemas used across organizations should be migrated to azure

> 6. Evaluate if the `fhir_mapping` files for HL7->FHIR library should move to azure

Reviewer: Only the end user translations should move?

Author: I kept this open ended since it was unclear to me what is an "end user" is going to be... I think that is mostly going to be in the `extension`, but this structure is more limited since we have to conform to the underlying library

    </details>
    <details>
      <summary>Staging</summary>

        ```
            /local-azure-container
                /fhir_transforms
                    valid-2023-04-02T23:53:00.488Z.txt
                    previous-valid-2023-03-29T23:53:00.488Z.txt
                    /default
                        input.fhir
                        output.fhir
                        default-sender-transform.yml
                    /Flexion
                        flexion-sender-transform.yml
                        input.fhir
                        output.fhir
        ```
    </details>
    <details>
      <summary>Production</summary>

        ```
            /local-azure-container
                /fhir_transforms
                    valid-2023-04-03T23:53:00.488Z.txt
                    previous-valid-2023-03-29T23:53:00.488Z.txt
                    /default
                        input.fhir
                        output.fhir
                        default-sender-transform.yml
                    /Flexion
                        flexion-sender-transform.yml
                        input.fhir
                        output.fhir
        ```
    </details>

#### Validation

> The initial validation will mimic how the current integration tests work.  Each translation schema will:

Reviewer 2: The current integration tests allow you test the whole integration from sender to receiver. It seems like this new set-up will only allow you test one piece of the integration(FHIR-> HL7, HL7->FHIR, and FHIR->FHIR). It would be missing the ability to test HL7 -> FHIR -> FHIR -> HL7 and FHIR -> FHIR -> HL7. This helps us ensure that making a change to a sender schema doesn't break the receiver schemas or identify any changes we need to make on the receiver side.

Author: Summarizing  @victor-chaparro and I's conversation

- Specifically discussed the simple report sender transform that subtracts 15 minutes, but this is only applicable for CA.  This is an example where a sender transform could break downstream receivers
- However, this is actually a current problem; there is a HL7->FHIR->FHIR-HL7 integration test for California that would catch any issues that a sender transform would cause for CA, but there is no check that an additional receiver had been setup to receive data
  - The integration tests mitigate this problem, but still leaves a gap
- The root of the problem is overall unclear guidelines on when a change should be an update to a sender transform or receiver transform
  - New ticket captures the work to create documentation that will explain: #10435

- be checked to make sure it can be successfully
- ran with its sample input and checked against its sample output failing translation schemas will be indicated in the output

> 

Reviewer 2: Merge these into one bullet point? They show as separate bullet points.
```suggestion
- be checked to make sure it can be successfully ran with its sample input and checked against its sample output 
- failing translation schemas will be indicated in the output
```

        ```
    </details>

#### Validation

The initial validation will mimic how the current integration tests work.  Each translation schema will:

- be checked to make sure it can be successfully
- ran with its sample input and checked against its sample output failing translation schemas will be indicated in the output

Some follow-up improvements could include indicating which lines did not match between the sample output and produced value
or what line was invalid in the schema.

### Creating/updating translation schemas used across organizations

These will continue to live in the source code and will only get updated as part of the application getting deployed.  Changes to these translation
schemas will need to be handled very carefully and there will be two primary update paths.

Non-breaking changes:

These can be made by directly editing the file and opening a PR to merge update.

Breaking changes:

A breaking change will need to be handled differently as there is no longer a guarantee that an update to a translation schema in the source

> code will be updated synchronously with files that live azure.  In order to support this use case a versioning system will be introduced

Reviewer: ```suggestion
code will be updated synchronously with files that live in azure.  In order to support this use case a versioning system will be introduced
```

> when required

Reviewer: ```suggestion

```

```
/hl7_mapping
    /ADT_A01
      ADT-A01-base.yml
      /v1
        ADT-A01-base.yml
```

In the example above, the `/v1` directory contains a breaking change in the translation schema.  A developer would merge and deploy a PR

> that introduces the new version and then separately update all receivers schemas that consume it.

Reviewer: ```suggestion
that introduces the new version and then separately update all receivers schemas that consume it using the normal azure translation update and validation process.
```

Reviewer: Or is this not right?

Author: Yes, that's correct
